Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp

```diff
@@ -0,0 +1,238 @@
+  auto *bindata = xbins->data();
+  HistogramData::BinEdges binedges(&bindata[0], &bindata[xbins->size()]);
+
+  API::MatrixWorkspace_sptr ws{DataObjects::create<DataObjects::Workspace2D>(
+      *m_workspace, nspectra, binedges)};
+
+  ws->setIndexInfo(m_workspace->indexInfo());
+  auto data = histoMsg->data_as_ArrayDouble()->value();
+
+  // Set the units
+  ws->getAxis(0)->setUnit(metadimx->unit()->c_str());
+  ws->setYUnit(metadimy->unit()->c_str());
+
+  std::vector<double> counts;
+  for (size_t i = 0; i < nspectra; ++i) {
+    const double *start = data->data() + (i * nbins);
+    counts.assign(start, start + nbins);
+    ws->setCounts(i, counts);
+  }
+
+  return ws;
+}
+
+/**
+ * Start decoding data from the streams into the internal buffers.
+ * Implementation designed to be entry point for new thread of execution.
+ * It catches all thrown exceptions.
+ */
+void KafkaHistoStreamDecoder::captureImpl() {
+  m_capturing = true;
+  try {
+    captureImplExcept();
+  } catch (std::exception &exc) {
+    m_exception = boost::make_shared<std::runtime_error>(exc.what());
+  } catch (...) {
+    m_exception = boost::make_shared<std::runtime_error>(
+        "KafkaEventStreamDecoder: Unknown exception type caught.");
+  }
+  m_capturing = false;
+}
+
+/**
+ * Exception-throwing variant of captureImpl(). Do not call this directly
+ */
+void KafkaHistoStreamDecoder::captureImplExcept() {
+  g_log.information("Event capture starting");
+
+  m_interrupt = false;
+  std::string buffer;
+  int64_t offset;
+  int32_t partition;
+  std::string topicName;
+
+  while (!m_interrupt) {
+    // Pull in events
+    m_histoStream->consumeMessage(&buffer, offset, partition, topicName);
+
+    // No events, wait for some to come along...
+    if (buffer.empty()) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+      continue;
+    }
+
+    // Lock so we don't overwrite buffer while workspace is being extracted
+    {
+      std::lock_guard<std::mutex> lock(m_buffer_mutex);
+      m_buffer = buffer;
+    }
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  }
+  g_log.debug("Histo capture finished");
+}
+
+DataObjects::Workspace2D_sptr KafkaHistoStreamDecoder::createBufferWorkspace() {
+  API::MatrixWorkspace_sptr workspace;
+
+  try {
+    auto alg = API::AlgorithmManager::Instance().createUnmanaged(
+        "LoadEmptyInstrument");
+    // Do not put the workspace in the ADS
+    alg->setChild(true);
+    alg->initialize();
+    alg->setPropertyValue("InstrumentName", m_instrumentName);
+    alg->setPropertyValue("OutputWorkspace", "ws");
+    alg->execute();
+    workspace = alg->getProperty("OutputWorkspace");
+  } catch (std::exception &exc) {
+    g_log.error() << "Error loading empty instrument '" << m_instrumentName
+                  << "': " << exc.what() << "\n";
+    throw;
+  }
+
+  return boost::dynamic_pointer_cast<DataObjects::Workspace2D>(workspace);
```

Comment:
Is this guranteed by the algorithm but not enforced by the property type for the return workspace? i.e. any chance this will fail?